File: src/typing/node_cache.ml
(*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

module Ast = Flow_ast
open Loc_collections

type cache = {
  annotations: (ALoc.t, ALoc.t * Type.t) Ast.Type.annotation ALocMap.t;
  expressions: (ALoc.t, ALoc.t * Type.t) Ast.Expression.t ALocMap.t;
  statements: (ALoc.t, ALoc.t * Type.t) Ast.Statement.t ALocMap.t;
  jsx_body:
    (Type.unresolved_param list * (ALoc.t * (ALoc.t, ALoc.t * Type.t) Ast.JSX.child list)) ALocMap.t;
  functions: (Type.t * (ALoc.t, ALoc.t * Type.t) Ast.Function.t) ALocMap.t;
  function_sigs:
    ( Func_class_sig_types.Func_stmt_sig_types.t
    * ((ALoc.t, ALoc.t * Type.t) Ast.Function.Params.t ->
      (ALoc.t, ALoc.t * Type.t) Ast.Function.body ->
      Type.t ->
      (ALoc.t, ALoc.t * Type.t) Ast.Function.t
      )
    )
    ALocMap.t;
  aliases: (Type.t * (ALoc.t, ALoc.t * Type.t) Ast.Statement.TypeAlias.t) ALocMap.t;
  opaques: (Type.t * (ALoc.t, ALoc.t * Type.t) Ast.Statement.OpaqueType.t) ALocMap.t;
  interfaces: (Type.t * (ALoc.t, ALoc.t * Type.t) Ast.Statement.Interface.t) ALocMap.t;
  declared_classes: (Type.t * (ALoc.t, ALoc.t * Type.t) Ast.Statement.DeclareClass.t) ALocMap.t;
  declared_components:
    (Type.t * (ALoc.t, ALoc.t * Type.t) Ast.Statement.DeclareComponent.t) ALocMap.t;
  declared_namespaces:
    (Type.t * (ALoc.t, ALoc.t * Type.t) Ast.Statement.DeclareNamespace.t) ALocMap.t;
  classes: (Type.t * (ALoc.t, ALoc.t * Type.t) Ast.Class.t) ALocMap.t;
  class_sigs:
    ( Type.t
    * Type.t
    * Func_class_sig_types.Class_stmt_sig_types.t
    * (Type.t -> (ALoc.t, ALoc.t * Type.t) Ast.Class.t)
    )
    ALocMap.t;
  tparams: ((ALoc.t, ALoc.t * Type.t) Ast.Type.TypeParam.t * Type.typeparam * Type.t) ALocMap.t;
  component_sigs:
    ( Component_sig_types.Component_declaration_sig_types.t
    * ((ALoc.t, ALoc.t * Type.t) Ast.Statement.ComponentDeclaration.Params.t ->
      ALoc.t * (ALoc.t, ALoc.t * Type.t) Ast.Statement.Block.t ->
      Type.t ->
      (ALoc.t, ALoc.t * Type.t) Ast.Statement.ComponentDeclaration.t
      )
    )
    ALocMap.t;
}

type t = cache ref

let mk_empty () =
  ref
    {
      annotations = ALocMap.empty;
      expressions = ALocMap.empty;
      statements = ALocMap.empty;
      jsx_body = ALocMap.empty;
      functions = ALocMap.empty;
      function_sigs = ALocMap.empty;
      aliases = ALocMap.empty;
      opaques = ALocMap.empty;
      interfaces = ALocMap.empty;
      declared_classes = ALocMap.empty;
      declared_components = ALocMap.empty;
      declared_namespaces = ALocMap.empty;
      classes = ALocMap.empty;
      class_sigs = ALocMap.empty;
      tparams = ALocMap.empty;
      component_sigs = ALocMap.empty;
    }

let set_annotation cache ((loc, _) as anno) =
  cache := { !cache with annotations = ALocMap.add loc anno !cache.annotations }

let set_expression cache (((loc, _), _) as exp) =
  cache := { !cache with expressions = ALocMap.add loc exp !cache.expressions }

let set_statement cache ((loc, _) as stmt) =
  cache := { !cache with statements = ALocMap.add loc stmt !cache.statements }

let set_jsx_children cache ((_, (loc, _)) as children) =
  cache := { !cache with jsx_body = ALocMap.add loc children !cache.jsx_body }

let set_function cache loc fn =
  cache := { !cache with functions = ALocMap.add loc fn !cache.functions }

let set_function_sig cache loc fn =
  cache := { !cache with function_sigs = ALocMap.add loc fn !cache.function_sigs }

let set_alias cache loc alias =
  cache := { !cache with aliases = ALocMap.add loc alias !cache.aliases }

let set_opaque cache loc opaque =
  cache := { !cache with opaques = ALocMap.add loc opaque !cache.opaques }

let set_interface cache loc inter =
  cache := { !cache with interfaces = ALocMap.add loc inter !cache.interfaces }

let set_declared_component cache loc component =
  cache :=
    { !cache with declared_components = ALocMap.add loc component !cache.declared_components }

let set_declared_class cache loc class_ =
  cache := { !cache with declared_classes = ALocMap.add loc class_ !cache.declared_classes }

let set_declared_namespace cache loc ns =
  cache := { !cache with declared_namespaces = ALocMap.add loc ns !cache.declared_namespaces }

let set_class cache loc class_ =
  cache := { !cache with classes = ALocMap.add loc class_ !cache.classes }

let set_class_sig cache loc class_ =
  cache := { !cache with class_sigs = ALocMap.add loc class_ !cache.class_sigs }

let set_tparam cache (((loc, _), _, _) as param) =
  cache := { !cache with tparams = ALocMap.add loc param !cache.tparams }

let set_component_sig cache loc c =
  cache := { !cache with component_sigs = ALocMap.add loc c !cache.component_sigs }

let get_annotation cache loc = ALocMap.find_opt loc !cache.annotations

let get_expression cache loc = ALocMap.find_opt loc !cache.expressions

let get_statement cache loc = ALocMap.find_opt loc !cache.statements

let get_jsx_children cache loc = ALocMap.find_opt loc !cache.jsx_body

let get_function_sig cache loc = ALocMap.find_opt loc !cache.function_sigs

let get_function cache loc = ALocMap.find_opt loc !cache.functions

let get_alias cache loc = ALocMap.find_opt loc !cache.aliases

let get_opaque cache loc = ALocMap.find_opt loc !cache.opaques

let get_interface cache loc = ALocMap.find_opt loc !cache.interfaces

let get_declared_class cache loc = ALocMap.find_opt loc !cache.declared_classes

let get_declared_component cache loc = ALocMap.find_opt loc !cache.declared_components

let get_declared_namespace cache loc = ALocMap.find_opt loc !cache.declared_namespaces

let get_class cache loc = ALocMap.find_opt loc !cache.classes

let get_class_sig cache loc = ALocMap.find_opt loc !cache.class_sigs

let get_tparam cache loc = ALocMap.find_opt loc !cache.tparams

let get_component_sig cache loc = ALocMap.find_opt loc !cache.component_sigs

File: rust_port/crates/flow_typing_loc_env/src/node_cache.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::cell::RefCell;
use std::rc::Rc;

use dupe::Dupe;
use flow_aloc::ALoc;
use flow_aloc::ALocMap;
use flow_parser::ast;
use flow_parser::ast::types::TypeParam as AstTypeParam;
use flow_typing_type::type_::Type;
use flow_typing_type::type_::TypeParam;
use flow_typing_type::type_::UnresolvedParam;

use crate::component_sig_types;
use crate::func_class_sig_types;
use crate::match_pattern_ir;

pub type DeferredTypeGuardCheck<'cx, CX> =
    Option<Rc<dyn Fn(&CX) -> Result<(), flow_utils_concurrency::job_error::JobError> + 'cx>>;

struct Cache<'cx, CX: Clone> {
    annotations: ALocMap<ast::types::Annotation<ALoc, (ALoc, Type)>>,
    expressions: ALocMap<ast::expression::Expression<ALoc, (ALoc, Type)>>,
    statements: ALocMap<ast::statement::Statement<ALoc, (ALoc, Type)>>,
    jsx_body: ALocMap<(
        Vec<UnresolvedParam>,
        (ALoc, Vec<ast::jsx::Child<ALoc, (ALoc, Type)>>),
    )>,
    functions: ALocMap<(Type, ast::function::Function<ALoc, (ALoc, Type)>)>,
    function_sigs: ALocMap<(
        func_class_sig_types::func::Func<func_class_sig_types::StmtConfigTypes>,
        func_class_sig_types::func::Reconstruct,
        DeferredTypeGuardCheck<'cx, CX>,
    )>,
    aliases: ALocMap<(Type, ast::statement::TypeAlias<ALoc, (ALoc, Type)>)>,
    opaques: ALocMap<(Type, ast::statement::OpaqueType<ALoc, (ALoc, Type)>)>,
    interfaces: ALocMap<(Type, ast::statement::Interface<ALoc, (ALoc, Type)>)>,
    declared_classes: ALocMap<(Type, ast::statement::DeclareClass<ALoc, (ALoc, Type)>)>,
    declared_components: ALocMap<(Type, ast::statement::DeclareComponent<ALoc, (ALoc, Type)>)>,
    declared_namespaces: ALocMap<(Type, ast::statement::DeclareNamespace<ALoc, (ALoc, Type)>)>,
    class_sigs: ALocMap<(
        Type,
        Type,
        func_class_sig_types::class::Class<func_class_sig_types::StmtConfigTypes>,
        Rc<
            dyn Fn(
                    &CX,
                    Type,
                ) -> Result<
                    ast::class::Class<ALoc, (ALoc, Type)>,
                    flow_utils_concurrency::job_error::JobError,
                > + 'cx,
        >,
    )>,
    record_sigs: ALocMap<(
        Type,
        Type,
        func_class_sig_types::class::Class<func_class_sig_types::StmtConfigTypes>,
        Rc<
            dyn Fn(
                    &CX,
                    Type,
                ) -> Result<
                    ast::statement::RecordDeclaration<ALoc, (ALoc, Type)>,
                    flow_utils_concurrency::job_error::JobError,
                > + 'cx,
        >,
    )>,
    tparams: ALocMap<(AstTypeParam<ALoc, (ALoc, Type)>, TypeParam, Type)>,
    component_sigs: ALocMap<(
        component_sig_types::component_sig::ComponentSig,
        component_sig_types::component_sig::ReconstructComponent,
    )>,
    match_patterns: ALocMap<(ast::match_pattern::MatchPattern<ALoc, (ALoc, Type)>, bool)>,
    match_pattern_value_unions: ALocMap<match_pattern_ir::value_union::ValueUnion<'cx, CX>>,
    // Intermediate PatternUnion state for incremental building
    match_pattern_unions: ALocMap<(match_pattern_ir::pattern_union::PatternUnion, i32)>,
}

pub struct NodeCache<'cx, CX: Clone>(Rc<RefCell<Cache<'cx, CX>>>);

impl<'cx, CX: Clone> NodeCache<'cx, CX> {
    /// Clear all cached entries, releasing any closures that may hold
    /// strong references to Context (breaking Rc cycles).
    pub fn clear(&self) {
        let mut cache = self.0.borrow_mut();
        cache.annotations = ALocMap::new();
        cache.expressions = ALocMap::new();
        cache.statements = ALocMap::new();
        cache.jsx_body = ALocMap::new();
        cache.functions = ALocMap::new();
        cache.function_sigs = ALocMap::new();
        cache.aliases = ALocMap::new();
        cache.opaques = ALocMap::new();
        cache.interfaces = ALocMap::new();
        cache.declared_classes = ALocMap::new();
        cache.declared_components = ALocMap::new();
        cache.declared_namespaces = ALocMap::new();
        cache.class_sigs = ALocMap::new();
        cache.record_sigs = ALocMap::new();
        cache.tparams = ALocMap::new();
        cache.component_sigs = ALocMap::new();
        cache.match_patterns = ALocMap::new();
        cache.match_pattern_value_unions = ALocMap::new();
        cache.match_pattern_unions = ALocMap::new();
    }

    pub fn mk_empty() -> NodeCache<'cx, CX> {
        NodeCache(Rc::new(RefCell::new(Cache {
            annotations: ALocMap::new(),
            expressions: ALocMap::new(),
            statements: ALocMap::new(),
            jsx_body: ALocMap::new(),
            functions: ALocMap::new(),
            function_sigs: ALocMap::new(),
            aliases: ALocMap::new(),
            opaques: ALocMap::new(),
            interfaces: ALocMap::new(),
            declared_classes: ALocMap::new(),
            declared_components: ALocMap::new(),
            declared_namespaces: ALocMap::new(),
            class_sigs: ALocMap::new(),
            record_sigs: ALocMap::new(),
            tparams: ALocMap::new(),
            component_sigs: ALocMap::new(),
            match_patterns: ALocMap::new(),
            match_pattern_value_unions: ALocMap::new(),
            match_pattern_unions: ALocMap::new(),
        })))
    }

    pub fn set_annotation(&self, anno: ast::types::Annotation<ALoc, (ALoc, Type)>) {
        let loc = anno.loc.dupe();
        self.0.borrow_mut().annotations.insert(loc, anno);
    }

    pub fn set_expression(&self, exp: ast::expression::Expression<ALoc, (ALoc, Type)>) {
        let (loc, _) = exp.loc();
        let loc = loc.clone();
        self.0.borrow_mut().expressions.insert(loc, exp);
    }

    pub fn set_statement(&self, stmt: ast::statement::Statement<ALoc, (ALoc, Type)>) {
        let loc = stmt.loc().clone();
        self.0.borrow_mut().statements.insert(loc, stmt);
    }

    pub fn set_jsx_children(
        &self,
        children: (
            Vec<UnresolvedParam>,
            (ALoc, Vec<ast::jsx::Child<ALoc, (ALoc, Type)>>),
        ),
    ) {
        let (_, (loc, _)) = &children;
        let loc = loc.clone();
        self.0.borrow_mut().jsx_body.insert(loc, children);
    }

    pub fn set_function(
        &self,
        loc: ALoc,
        func: (Type, ast::function::Function<ALoc, (ALoc, Type)>),
    ) {
        self.0.borrow_mut().functions.insert(loc, func);
    }

    pub fn set_function_sig(
        &self,
        loc: ALoc,
        func_sig: (
            func_class_sig_types::func::Func<func_class_sig_types::StmtConfigTypes>,
            func_class_sig_types::func::Reconstruct,
            DeferredTypeGuardCheck<'cx, CX>,
        ),
    ) {
        self.0.borrow_mut().function_sigs.insert(loc, func_sig);
    }

    pub fn set_alias(
        &self,
        loc: ALoc,
        alias: (Type, ast::statement::TypeAlias<ALoc, (ALoc, Type)>),
    ) {
        self.0.borrow_mut().aliases.insert(loc, alias);
    }

    pub fn set_opaque(
        &self,
        loc: ALoc,
        opaque: (Type, ast::statement::OpaqueType<ALoc, (ALoc, Type)>),
    ) {
        self.0.borrow_mut().opaques.insert(loc, opaque);
    }

    pub fn set_interface(
        &self,
        loc: ALoc,
        inter: (Type, ast::statement::Interface<ALoc, (ALoc, Type)>),
    ) {
        self.0.borrow_mut().interfaces.insert(loc, inter);
    }

    pub fn set_declared_component(
        &self,
        loc: ALoc,
        component: (Type, ast::statement::DeclareComponent<ALoc, (ALoc, Type)>),
    ) {
        self.0
            .borrow_mut()
            .declared_components
            .insert(loc, component);
    }

    pub fn set_declared_class(
        &self,
        loc: ALoc,
        class_: (Type, ast::statement::DeclareClass<ALoc, (ALoc, Type)>),
    ) {
        self.0.borrow_mut().declared_classes.insert(loc, class_);
    }

    pub fn set_declared_namespace(
        &self,
        loc: ALoc,
        ns: (Type, ast::statement::DeclareNamespace<ALoc, (ALoc, Type)>),
    ) {
        self.0.borrow_mut().declared_namespaces.insert(loc, ns);
    }

    pub fn set_class_sig(
        &self,
        loc: ALoc,
        class_: (
            Type,
            Type,
            func_class_sig_types::class::Class<func_class_sig_types::StmtConfigTypes>,
            Rc<
                dyn Fn(
                        &CX,
                        Type,
                    ) -> Result<
                        ast::class::Class<ALoc, (ALoc, Type)>,
                        flow_utils_concurrency::job_error::JobError,
                    > + 'cx,
            >,
        ),
    ) {
        self.0.borrow_mut().class_sigs.insert(loc, class_);
    }

    pub fn set_record_sig(
        &self,
        loc: ALoc,
        record: (
            Type,
            Type,
            func_class_sig_types::class::Class<func_class_sig_types::StmtConfigTypes>,
            Rc<
                dyn Fn(
                        &CX,
                        Type,
                    ) -> Result<
                        ast::statement::RecordDeclaration<ALoc, (ALoc, Type)>,
                        flow_utils_concurrency::job_error::JobError,
                    > + 'cx,
            >,
        ),
    ) {
        self.0.borrow_mut().record_sigs.insert(loc, record);
    }

    pub fn set_tparam(&self, param: (AstTypeParam<ALoc, (ALoc, Type)>, TypeParam, Type)) {
        let (ast_tparam, _, _) = &param;
        let loc = ast_tparam.loc.clone();
        self.0.borrow_mut().tparams.insert(loc, param);
    }

    pub fn set_component_sig(
        &self,
        loc: ALoc,
        c: (
            component_sig_types::component_sig::ComponentSig,
            component_sig_types::component_sig::ReconstructComponent,
        ),
    ) {
        self.0.borrow_mut().component_sigs.insert(loc, c);
    }

    pub fn set_match_pattern(
        &self,
        loc: ALoc,
        p: (ast::match_pattern::MatchPattern<ALoc, (ALoc, Type)>, bool),
    ) {
        self.0.borrow_mut().match_patterns.insert(loc, p);
    }

    pub fn set_match_pattern_value_union(
        &self,
        loc: ALoc,
        v: match_pattern_ir::value_union::ValueUnion<'cx, CX>,
    ) {
        self.0
            .borrow_mut()
            .match_pattern_value_unions
            .insert(loc, v);
    }

    pub fn set_match_pattern_union(
        &self,
        loc: ALoc,
        v: (match_pattern_ir::pattern_union::PatternUnion, i32),
    ) {
        self.0.borrow_mut().match_pattern_unions.insert(loc, v);
    }

    pub fn get_annotation(&self, loc: &ALoc) -> Option<ast::types::Annotation<ALoc, (ALoc, Type)>> {
        self.0.borrow().annotations.get(loc).cloned()
    }

    pub fn get_expression(
        &self,
        loc: &ALoc,
    ) -> Option<ast::expression::Expression<ALoc, (ALoc, Type)>> {
        self.0.borrow().expressions.get(loc).cloned()
    }

    pub fn get_statement(
        &self,
        loc: &ALoc,
    ) -> Option<ast::statement::Statement<ALoc, (ALoc, Type)>> {
        self.0.borrow().statements.get(loc).cloned()
    }

    pub fn get_jsx_children(
        &self,
        loc: &ALoc,
    ) -> Option<(
        Vec<UnresolvedParam>,
        (ALoc, Vec<ast::jsx::Child<ALoc, (ALoc, Type)>>),
    )> {
        self.0.borrow().jsx_body.get(loc).cloned()
    }

    pub fn get_function_sig(
        &self,
        loc: &ALoc,
    ) -> Option<(
        func_class_sig_types::func::Func<func_class_sig_types::StmtConfigTypes>,
        func_class_sig_types::func::Reconstruct,
        DeferredTypeGuardCheck<'cx, CX>,
    )> {
        self.0.borrow().function_sigs.get(loc).cloned()
    }

    pub fn get_function(
        &self,
        loc: &ALoc,
    ) -> Option<(Type, ast::function::Function<ALoc, (ALoc, Type)>)> {
        self.0.borrow().functions.get(loc).cloned()
    }

    pub fn get_alias(
        &self,
        loc: &ALoc,
    ) -> Option<(Type, ast::statement::TypeAlias<ALoc, (ALoc, Type)>)> {
        self.0.borrow().aliases.get(loc).cloned()
    }

    pub fn get_opaque(
        &self,
        loc: &ALoc,
    ) -> Option<(Type, ast::statement::OpaqueType<ALoc, (ALoc, Type)>)> {
        self.0.borrow().opaques.get(loc).cloned()
    }

    pub fn get_interface(
        &self,
        loc: &ALoc,
    ) -> Option<(Type, ast::statement::Interface<ALoc, (ALoc, Type)>)> {
        self.0.borrow().interfaces.get(loc).cloned()
    }

    pub fn get_declared_class(
        &self,
        loc: &ALoc,
    ) -> Option<(Type, ast::statement::DeclareClass<ALoc, (ALoc, Type)>)> {
        self.0.borrow().declared_classes.get(loc).cloned()
    }

    pub fn get_declared_component(
        &self,
        loc: &ALoc,
    ) -> Option<(Type, ast::statement::DeclareComponent<ALoc, (ALoc, Type)>)> {
        self.0.borrow().declared_components.get(loc).cloned()
    }

    pub fn get_declared_namespace(
        &self,
        loc: &ALoc,
    ) -> Option<(Type, ast::statement::DeclareNamespace<ALoc, (ALoc, Type)>)> {
        self.0.borrow().declared_namespaces.get(loc).cloned()
    }

    pub fn get_class_sig(
        &self,
        loc: &ALoc,
    ) -> Option<(
        Type,
        Type,
        func_class_sig_types::class::Class<func_class_sig_types::StmtConfigTypes>,
        Rc<
            dyn Fn(
                    &CX,
                    Type,
                ) -> Result<
                    ast::class::Class<ALoc, (ALoc, Type)>,
                    flow_utils_concurrency::job_error::JobError,
                > + 'cx,
        >,
    )> {
        self.0.borrow().class_sigs.get(loc).cloned()
    }

    pub fn get_record_sig(
        &self,
        loc: &ALoc,
    ) -> Option<(
        Type,
        Type,
        func_class_sig_types::class::Class<func_class_sig_types::StmtConfigTypes>,
        Rc<
            dyn Fn(
                    &CX,
                    Type,
                ) -> Result<
                    ast::statement::RecordDeclaration<ALoc, (ALoc, Type)>,
                    flow_utils_concurrency::job_error::JobError,
                > + 'cx,
        >,
    )> {
        self.0.borrow().record_sigs.get(loc).cloned()
    }

    pub fn get_tparam(
        &self,
        loc: &ALoc,
    ) -> Option<(AstTypeParam<ALoc, (ALoc, Type)>, TypeParam, Type)> {
        self.0.borrow().tparams.get(loc).cloned()
    }

    pub fn get_component_sig(
        &self,
        loc: &ALoc,
    ) -> Option<(
        component_sig_types::component_sig::ComponentSig,
        component_sig_types::component_sig::ReconstructComponent,
    )> {
        self.0.borrow().component_sigs.get(loc).cloned()
    }

    pub fn get_match_pattern(
        &self,
        loc: &ALoc,
    ) -> Option<(ast::match_pattern::MatchPattern<ALoc, (ALoc, Type)>, bool)> {
        self.0.borrow().match_patterns.get(loc).cloned()
    }

    pub fn get_match_pattern_value_union(
        &self,
        loc: &ALoc,
    ) -> Option<match_pattern_ir::value_union::ValueUnion<'cx, CX>> {
        self.0.borrow().match_pattern_value_unions.get(loc).cloned()
    }

    pub fn get_match_pattern_union(
        &self,
        loc: &ALoc,
    ) -> Option<(match_pattern_ir::pattern_union::PatternUnion, i32)> {
        self.0.borrow().match_pattern_unions.get(loc).cloned()
    }
}
